src/Panel.js

```diff
@@ -1,236 +1,143 @@
+    expanded: PropTypes.bool,
+  })
+};
 
-    if (e.selected) {
-      this.setState({ expanded: !this.state.expanded });
-    }
+class Panel extends React.Component {
+  constructor(...args) {
+    super(...args);
+    this.handleToggle = this.handleToggle.bind(this);
   }
 
-  renderHeader(collapsible, header, id, role, expanded, bsProps) {
-    const titleClassName = prefix(bsProps, 'title');
+  getChildContext() {
+    const { eventKey, id } = this.props;
+    const { getId = defaultGetId } = this.context.$bs_panelGroup || {};
 
-    if (!collapsible) {
-      if (!React.isValidElement(header)) {
-        return header;
-      }
-
-      return cloneElement(header, {
-        className: classNames(header.props.className, titleClassName),
+    let getIds = null;
+    if (getId || id) {
+      getIds = () => ({
+        headingId: getId(eventKey || id, 'HEADING'),
+        collapseId: getId(eventKey || id, 'COLLAPSE')
       });
     }
 
-    if (!React.isValidElement(header)) {
-      return (
-        <h4 role="presentation" className={titleClassName}>
-          {this.renderAnchor(header, id, role, expanded)}
-        </h4>
-      );
-    }
-
-    return cloneElement(header, {
-      className: classNames(header.props.className, titleClassName),
-      children: this.renderAnchor(header.props.children, id, role, expanded),
-    });
-  }
-
-  renderAnchor(header, id, role, expanded) {
-    return (
-      <a
-        role={role}
-        href={id && `#${id}`}
-        onClick={this.handleClickTitle}
-        aria-controls={id}
-        aria-expanded={expanded}
-        aria-selected={expanded}
-        className={expanded ? null : 'collapsed' }
-      >
-        {header}
-      </a>
-    );
-  }
-
-  renderCollapsibleBody(
-    id, expanded, role, children, bsProps, animationHooks
-  ) {
-    return (
-      <Collapse in={expanded} {...animationHooks}>
-        <div
-          id={id}
-          role={role}
-          className={prefix(bsProps, 'collapse')}
-          aria-hidden={!expanded}
-        >
-          {this.renderBody(children, bsProps)}
-        </div>
-      </Collapse>
-    );
-  }
-
-  renderBody(rawChildren, bsProps) {
-    const children = [];
-    let bodyChildren = [];
-
-    const bodyClassName = prefix(bsProps, 'body');
-
-    function maybeAddBody() {
-      if (!bodyChildren.length) {
-        return;
+    return {
+      $bs_panel: {
+        getIds,
+        bsClass: this.props.bsClass,
+        expanded: this.getExpanded(),
+        onToggle: this.handleToggle,
       }
+    };
+  }
 
-      // Derive the key from the index here, since we need to make one up.
-      children.push(
-        <div key={children.length} className={bodyClassName}>
-          {bodyChildren}
-        </div>
-      );
+  handleToggle(e) {
+    const { onToggle } = this.context.$bs_panelGroup || {};
+    const expanded = !this.getExpanded();
 
-      bodyChildren = [];
+    this.props.onToggle(expanded, e);
+    if (onToggle) {
```

Comment:
I don't think so this is controllable so you can just not flush the value as the preventDefault, as we do elsewhere i believe